### PR TITLE
Update to buildroot 2023.02.9 & Golang 1.21.6

### PR DESCRIPTION
**Pull request by travier**
Heavily inspired by https://github.com/kubernetes/minikube/pull/17093

Building and launching an instance using the KVM2 driver & cri-o runtime works.

Did not do any further tests so far.

Fixes: https://github.com/kubernetes/minikube/issues/15841

### CODE REVIEW

**Comment by afbjorklund** [comment]
You might want to separate the ISO updates from the CI updates

**Comment by travier** [comment]
They were all together in the parent PRs. A lot of the ISO updates are removing packages to use the one from the new buildroot version.

Could you clarify which changes do you think I should split?

**Comment by spowelljr** [comment]
The hack/update change is because we were managing the CNI plugins package ourself, but the new Buildroot version has it included, so I was removing our CI from auto updating the CNI plugins file since the ISO no longer exist, so I think it makes sense to include it

**Comment by spowelljr** [comment]
ok-to-build-iso

**Comment by travier** [comment]
- cni-pluings from buildroot does not include the `host-local` plugin
- crun version in buildroot is only `1.7.2`

**Comment by spowelljr** [comment]
Just a heads up that we're resolving issues with our CI infrastructure so won't be able to build the ISO right now, but bear with us and we'll try building this as soon as we can

**Comment by medyagh** [comment]
@travier thank you very much for this contribution, glad to find another developer touching this code, this part of the code requires special experties, our CI machine is currently having issue, as soon as we fix that we can build the ISO again

**Comment by travier** [comment]
How would you prefer to go for crun & cni-plugins? Should we keep the updated/custom versions in minikube or should we patch the buildroot ones?

From experiments so far, we can not have external packages with the same names as the ones in buildroot so we would need to rename them to keep them in the minikube tree. Otherwise this will need "large" patches for the buildroot tree.

**Comment by spowelljr** [comment]
From a maintenance standpoint I think having a custom module with a different name would be ideal. Unless there is an easy to to update the patches. I'm just thinking if Buildroot updates crun they we have to remove the patches that are now included, and a new version comes out we have to add new patches, seems like more work than the custom package.

**Comment by travier** [comment]
So, with this update, this still runs crun 1.7.2 (not sure why it's not pulling the latest yet) but it uses the right cni-plugins and thus the pods can run.

I pulled in some other package updates. Let me know if you prefer to have them split from this PR.

**Comment by spowelljr** [comment]
Awesome thanks! I've made significant progress in resolving the CI infrastructure issues. The machines are running again, but there are 1,000 test jobs in the backlog, I'm working on resolving the login issue and then I can clear the backlog and we can try building this.

**Comment by spowelljr** [comment]
Sorry for the delay, just finished resolving the CI issues, let's trying building this finally

**Comment by spowelljr** [comment]
> I pulled in some other package updates. Let me know if you prefer to have them split from this PR.

Due to the backup in the whole CI I'm happy to merge this with all the package updates as long as the build is successful

**Comment by travier** [comment]
For the GitHub workflow, this fails on:
```
---------------- 1 Failures :( ----------------------------
[
  "TestFunctional/parallel/DashboardCmd"
]
```

I don't have access to the ISO PR Build Push logs.

**Comment by spowelljr** [comment]
The DashboardCmd failure you can ignore, the test is flaky and needs to be fixed.

The ISO build failed with VCS stamping cri-dockerd
```
16:44:34 mkdir -p build/linux/cri-dockerd
16:44:34 cd /home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/cri-dockerd-aarch64-b58acf8f78f9d7bce1241d1cddb0932e7101f278 && env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -trimpath -ldflags "-s -w -buildid=b58acf8 -X github.com/Mirantis/cri-dockerd/cmd/version.Version=0.3.3 -X github.com/Mirantis/cri-dockerd/cmd/version.PreRelease=`grep -q dev <<< "0.3.3" && echo "pre" || echo ""` -X github.com/Mirantis/cri-dockerd/cmd/version.GitCommit=b58acf8" -o cri-dockerd
16:44:35 error obtaining VCS status: exit status 128
16:44:35 	Use -buildvcs=false to disable VCS stamping.
```

**Comment by travier** [comment]
Updated to fix the issue above. It builds on x86_64 for me. Not tested.

**Comment by travier** [comment]
https://github.com/kubernetes/minikube/actions/runs/7914018515/job/21615627575:
```
functional_virtualbox_macos : completed with 0 / 0 failures in 35 min 1 seconds
*******************0 Passes :) *******************
null
*******************************************************
----------------0 Failures :(----------------------------
null
-------------------------------------------------------
*** 0 Passed! ***
```

🤔

**Comment by spowelljr** [comment]
VirtualBox is flaky and sometimes doesn't run the tests as well, you can ignore it

**Comment by spowelljr** [comment]
Huh, still getting VCS error on aarch64, also it's still showing `amd64`, did it no pick up the changes?
```
09:59:15 /usr/bin/make APP_DIR=/home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/cri-dockerd-aarch64-b58acf8f78f9d7bce1241d1cddb0932e7101f278 -C /home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/cri-dockerd-aarch64-b58acf8f78f9d7bce1241d1cddb0932e7101f278/packaging static
09:59:15 for p in static-linux cross-mac cross-win cross-arm; do \
09:59:15 	/usr/bin/make -C static APP_DIR=/home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/cri-dockerd-aarch64-b58acf8f78f9d7bce1241d1cddb0932e7101f278 GO_VERSION=1.21.6 ${p} || exit 1; \
09:59:15 done
09:59:15 mkdir -p build/linux/cri-dockerd
09:59:15 cd /home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/cri-dockerd-aarch64-b58acf8f78f9d7bce1241d1cddb0932e7101f278 && env CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -trimpath -ldflags "-s -w -buildid=b58acf8 -X github.com/Mirantis/cri-dockerd/cmd/version.Version=0.3.3 -X github.com/Mirantis/cri-dockerd/cmd/version.PreRelease=`grep -q dev <<< "0.3.3" && echo "pre" || echo ""` -X github.com/Mirantis/cri-dockerd/cmd/version.GitCommit=b58acf8" -o cri-dockerd
09:59:16 error obtaining VCS status: exit status 128
09:59:16 	Use -buildvcs=false to disable VCS stamping.
09:59:16 make[5]: *** [Makefile:52: static-linux] Error 1
09:59:16 make[4]: *** [Makefile:29: static] Error 1
09:59:16 make[3]: *** [Makefile:40: static] Error 2
09:59:16 make[2]: *** [package/pkg-generic.mk:283: /home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/cri-dockerd-aarch64-b58acf8f78f9d7bce1241d1cddb0932e7101f278/.stamp_built] Error 2
```

**Comment by spowelljr** [comment]
ok-to-build-iso

File: deploy/iso/minikube-iso/arch/aarch64/package/cri-dockerd-aarch64/cri-dockerd.mk [hunk 1]
```diff
@@ -23,6 +23,15 @@ CRI_DOCKERD_AARCH64_ENV = \
...
 CRI_DOCKERD_AARCH64_BUILDFLAGS = "-ldflags '-X github.com/Mirantis/cri-dockerd/version.Version=$(CRI_DOCKERD_AARCH64_VER) -X github.com/Mirantis/cri-dockerd/version.GitCommit=$(CRI_DOCKERD_AARCH64_REV)'"
 
+define CRI_DOCKERD_DISABLE_BUILDVCS
+	# Set -buildvcs=false to disable VCS stamping (fails in buildroot)
+	sed -i 's|go build |go build -buildvcs=false |' -i $(@D)/packaging/static/Makefile
+	# Use the GOARCH environment variable that we set
+	sed -i 's|GOARCH=$(ARCH) go build|GOARCH=$(GOARCH) go build|' -i $(@D)/packaging/static/Makefile
+endef
+
+CRI_DOCKERD_POST_EXTRACT_HOOKS += CRI_DOCKERD_DISABLE_BUILDVCS
```
Comment (spowelljr) [inline deploy/iso/minikube-iso/arch/aarch64/package/cri-dockerd-aarch64/cri-dockerd.mk:33]:
```suggestion
define CRI_DOCKERD_AARCH64_DISABLE_BUILDVCS
	# Set -buildvcs=false to disable VCS stamping (fails in buildroot)
	sed -i 's|go build |go build -buildvcs=false |' -i $(@D)/packaging/static/Makefile
	# Use the GOARCH environment variable that we set
	sed -i 's|GOARCH=$(ARCH) go build|GOARCH=$(GOARCH) go build|' -i $(@D)/packaging/static/Makefile
endef

CRI_DOCKERD_AARCH64_POST_EXTRACT_HOOKS += CRI_DOCKERD_DISABLE_BUILDVCS
```

**Review by spowelljr — CHANGES_REQUESTED** [review]
I think I found why, needs to `_AARCH64` to be picked up I believe

File: deploy/iso/minikube-iso/arch/aarch64/package/cri-dockerd-aarch64/cri-dockerd.mk (hunk 1 above)
Comment (travier) [reply deploy/iso/minikube-iso/arch/aarch64/package/cri-dockerd-aarch64/cri-dockerd.mk]: Good catch!

**Comment by spowelljr** [comment]
Error is:
```
17:25:19 >>>   Executing post-image script /home/jenkins/go/src/k8s.io/minikube/deploy/iso/minikube-iso/board/minikube/aarch64/post-image.sh
17:25:19 INFO: cmd: "mkdir -p "/home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/genimage.tmp"" (stderr):
17:25:19 INFO: cmd: "rm -rf "/home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/genimage.tmp"/*" (stderr):
17:25:19 INFO: cmd: "mkdir -p "/home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/genimage.tmp"" (stderr):
17:25:19 INFO: cmd: "cp -a "/tmp/tmp.RLV1aQfb2d" "/home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/build/genimage.tmp/root"" (stderr):
17:25:19 ERROR: file(efi-part/startup.nsh): stat(/home/jenkins/go/src/k8s.io/minikube/out/buildroot/output-aarch64/images/efi-part/startup.nsh) failed: No such file or directory
17:25:19 ERROR: vfat(efiboot.img): could not setup efi-part/startup.nsh
17:25:19 make[2]: *** [Makefile:816: target-post-image] Error 1
17:25:19 make[1]: *** [Makefile:82: _all] Error 2
17:25:19 make[1]: Leaving directory '/home/jenkins/go/src/k8s.io/minikube/out/buildroot'
17:25:19 make: *** [Makefile:301: minikube-iso-aarch64] Error 2
```

**Comment by spowelljr** [comment]
Looking at our current version of Buildroot: https://github.com/buildroot/buildroot/blob/2021.02.12/board/pc/genimage-efi.cfg

Comparing to this new version: https://github.com/buildroot/buildroot/blob/2023.02.9/board/pc/genimage-efi.cfg

The reference to:
```
    file startup.nsh {
      image = "efi-part/startup.nsh"
    }
```
Is gone

Here's the commit that removed it that has information on why it was removed: https://github.com/buildroot/buildroot/commit/3efb5e31fc05705ce3c46b1f0ec031978a5cfab6

So I think we need to remove it from `deploy/iso/minikube-iso/board/minikube/aarch64/genimage.cfg`

**Comment by travier** [comment]
Thanks! Updated.

**Comment by spowelljr** [comment]
ok-to-build-iso

**Comment by travier** [comment]
I've started looking at `TestStartStop/group/old-k8s-version/serial/DeployApp` and it's trying to run Kubernetes 1.16 with the latest cri-o, which is likely not going to work as cri-o is "version"-locked with kubernetes.

I have not looked at the other failures yet.

**Comment by spowelljr** [comment]
We talked about it in office hours today, basically as of 1.28(?) the cri-o is built for the matching version of Kubernetes and there is no backwards compatibility guaranteed. So the updated cri-o is not working with the old version of Kubernetes we're testing with (1.16). As long as it works with a newer version say 1.24 then I'm happy to bump the "old" version we use for testing.

**Comment by spowelljr** [comment]
This worked down to K8s v1.20 for me, so looks good to me

**Review by spowelljr — APPROVED** [review]

**Comment by spowelljr** [comment]
Thank you very much for your continued pursuit of this PR @travier, happy to be back on the current version of Buildroot, and unblocks using Go 1.21 and will allow us to install new packages that were blocked like Docker buildx

**Comment by travier** [comment]
@spowelljr Thanks for the help and reviews! I'll likely be looking at crun and other components updates now.

**Pull request merged**